src/services/kratos/auth-device-account.ts

```diff
@@ -0,0 +1,249 @@
+    deviceId: DeviceId
+  }): Promise<LoginWithDeviceSchemaResponse | KratosError> => {
+    const identifier = deviceId
+    const method = "password"
+
+    try {
+      const flow = await kratosPublic.createNativeLoginFlow()
+      const result = await kratosPublic.updateLoginFlow({
+        flow: flow.data.id,
+        updateLoginFlowBody: {
+          identifier,
+          method,
+          password,
+          traits: {
+            device_id: identifier,
+          },
+        },
+      })
+      const sessionToken = result.data.session_token as SessionToken
+
+      // note: this only works when whoami: required_aal = aal1
+      const kratosUserId = result.data.session.identity.id as UserId
+
+      return { sessionToken, kratosUserId }
+    } catch (err) {
+      if (err.message === "Request failed with status code 400") {
+        return new LikelyNoUserWithThisPhoneExistError(err.message || err)
+      }
+
+      if (err.message === "Request failed with status code 401") {
+        return new AuthenticationKratosError(err.message || err)
+      }
+
+      return new UnknownKratosError(err.message || err)
+    }
+  }
+
+  const createDeviceIdentity = async ({
+    deviceId,
+  }: {
+    deviceId: DeviceId
+  }): Promise<CreateKratosUserForDeviceSchemaResponse | KratosError> => {
+    const traits = { device: deviceId }
+    const identityBody: CreateIdentityBody = {
+      credentials: { password: { config: { password } } },
+      state: "active",
+      schema_id: "device_account_v0",
+      traits,
+    }
+    try {
+      const identityResults = await kratosAdmin.createIdentity({
+        createIdentityBody: identityBody,
+      })
+      console.log(identityResults.status)
+      const result = await loginDeviceAccount({ deviceId })
+      if (result instanceof KratosError) return result
+      const sessionToken = result.sessionToken
+      const kratosUserId = result.kratosUserId as UserId
+
+      const levelZeroAccountsConfig = getDefaultAccountsConfig()
+      levelZeroAccountsConfig.initialLevel = AccountLevel.Zero
+      const account = await createAccountForDeviceAccount({
+        userId: kratosUserId,
+        config: levelZeroAccountsConfig,
+        device: deviceId,
+      })
+      if (account instanceof Error) return account
+      return { sessionToken, kratosUserId }
+    } catch (err) {
+      if (err.message === "Request failed with status code 400") {
+        return new LikelyUserAlreadyExistError(err.message || err)
+      }
+
+      return new UnknownKratosError(err.message || err)
+    }
+  }
+
+  const upgradeToPhoneSchema = async ({
+    kratosUserId,
+    phone,
+  }: {
+    kratosUserId: UserId
+    phone: PhoneNumber
+  }) => {
+    let identity: Identity
+
+    try {
+      ;({ data: identity } = await kratosAdmin.getIdentity({ id: kratosUserId }))
+    } catch (err) {
+      if (err.message === "Request failed with status code 400") {
+        return new LikelyUserAlreadyExistError(err)
+      }
```

Comment:
not sure this error message make most sense here, we expect the user to already exist?

Reply:
yep. removed that error